File: engine/src/com/magma/engine/ui/RetroButtonGrid.java
package com.magma.engine.ui;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.Input.Keys;
import com.badlogic.gdx.graphics.Color;
import com.badlogic.gdx.graphics.g2d.Batch;
import com.badlogic.gdx.math.GridPoint2;
import com.badlogic.gdx.math.MathUtils;
import com.badlogic.gdx.scenes.scene2d.ui.Table;
import com.magma.engine.assets.Shapes;

import space.earlygrey.shapedrawer.ShapeDrawer;

public abstract class RetroButtonGrid extends Table {

    private final String[] options;
    private final RetroButton[][] buttons;
    private final int columns;
    private final int rows;
    private final GridPoint2 selected;

    private boolean hasFocus;
    private Color bgColor;
    private Color fgColor;

    public RetroButtonGrid(String... options) {
        this.options = options;
        this.columns = (int) Math.max(2, Math.sqrt(options.length));
        this.rows = options.length / columns;
        this.buttons = new RetroButton[columns][rows];
        this.selected = new GridPoint2(0, 0);

        for (int i = 0; i < options.length; i++) {
            final int id = i; // java lol
            String option = options[i];
            RetroButton butt = new RetroButton(option);
            if (i == 0) {
                butt.select();
            } else if (i % columns == 0) {
                row();
            }
            add(butt).expand();
            int x = i % columns;
            int y = i / columns;
            buttons[x][y] = butt;
        }

        this.setColor(Color.BLACK, Color.WHITE);
    }

    public void focus() {
        unfocus(); // clear previous user input
        hasFocus = true;
        // select the first button
        setCursor(0, 0);
    }

    public void unfocus() {
        hasFocus = false;
        deselectAll();
    }

    private void deselectAll() {
        // deselect all buttons
        for (int y = 0; y < rows; y++) {
            for (int x = 0; x < columns; x++) {
                buttons[x][y].deselect();
            }
        }
    }

    public void moveCursor(int x, int y) {
        setCursor(selected.x + x, selected.y + y);
    }

    public void setCursor(int x, int y) {
        selected.x = MathUtils.clamp(x, 0, columns - 1);
        selected.y = MathUtils.clamp(y, 0, rows - 1);
        deselectAll();
        buttons[selected.x][selected.y].select();
    }

    public void setColor(Color bgColor, Color fgColor) {
        this.bgColor = bgColor;
        this.fgColor = fgColor;
        for (int y = 0; y < buttons.length; y++) {
            for (int x = 0; x < buttons[0].length; x++) {
                buttons[x][y].setColor(fgColor);
            }
        }
    }

    @Override
    public void act(float delta) {
        if (hasFocus) {
            // TODO: use keymapping
            if (Gdx.input.isKeyJustPressed(Keys.ENTER)) {
                int id = getSelectedID();
                pressed(id);
            }
            if (Gdx.input.isKeyJustPressed(Keys.LEFT)) {
                moveCursor(-1, 0);
            }
            if (Gdx.input.isKeyJustPressed(Keys.RIGHT)) {
                moveCursor(1, 0);
            }
            if (Gdx.input.isKeyJustPressed(Keys.UP)) {
                moveCursor(0, -1);
            }
            if (Gdx.input.isKeyJustPressed(Keys.DOWN)) {
                moveCursor(0, 1);
            }
        }
        super.act(delta);
    }

    @Override
    public void draw(Batch batch, float parentAlpha) {
        ShapeDrawer sh = Shapes.getInstance();
        sh.setColor(bgColor);
        sh.filledRectangle(getX(), getY(), getWidth(), getHeight());
        sh.setColor(fgColor);
        sh.rectangle(getX(), getY(), getWidth(), getHeight(), 3f);
        super.draw(batch, parentAlpha);
    }

    protected abstract void pressed(int id);

    public RetroButton getSelected() {
        return buttons[selected.x][selected.y];
    }

    private int getSelectedID() {
        return selected.y * columns + selected.x;
    }

}

File: engine/src/com/magma/engine/ui/RetroButtonGrid.kt
package com.magma.engine.ui

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.Input
import com.badlogic.gdx.graphics.Color
import com.badlogic.gdx.graphics.g2d.Batch
import com.badlogic.gdx.math.MathUtils
import com.badlogic.gdx.scenes.scene2d.ui.Table
import com.magma.engine.assets.Shapes
import space.earlygrey.shapedrawer.ShapeDrawer
import kotlin.math.max
import kotlin.math.sqrt

abstract class RetroButtonGrid(vararg options: String) : Table() {
    private val options: Array<String>
    private val buttons: Array<RetroButton>
    private val gridColumns: Int
    private val gridRows: Int

    private var hasFocus = false
    private var selectedIndex: Int = 0
    private var bgColor: Color = Color.BLACK
    private var fgColor: Color = Color.WHITE

    init {
        this.options = arrayOf(*options)
        this.gridColumns = max(2.0, sqrt(options.size.toDouble())).toInt()
        this.gridRows = options.size / gridColumns
        buttons = Array(gridColumns * gridRows) { i ->
            val option = options[i]
            val butt = RetroButton(option)
            if (i == 0) {
                butt.select()
            } else if (i % gridColumns == 0) {
                row()
            }
            add(butt).expand()
            butt
        }
    }

    fun focus() {
        unfocus() // clear previous user input
        hasFocus = true
        // select the first button
        setCursor(0, 0)
    }

    fun unfocus() {
        hasFocus = false
        deselectAll()
    }

    private fun deselectAll() {
        // deselect all buttons
        for (button in buttons) {
            button.deselect()
        }
    }

    var cursorLocation: Int
        get() = selectedIndex
        set(value) {
            selectedIndex = MathUtils.clamp(value, 0, buttons.size - 1)
            deselectAll()
            buttons[selectedIndex].select()
        }


    fun moveCursor(x: Int, y: Int) {
        selectedIndex += gridColumns * y
        val curX = selectedIndex % gridColumns
        val curY = selectedIndex / gridRows
        setCursor(curX + x, curY + y)
    }

    fun setCursor(x: Int, y: Int) {
        cursorLocation = y * gridColumns + x
    }

    fun setColor(bgColor: Color, fgColor: Color) {
        this.bgColor = bgColor
        this.fgColor = fgColor
        for (button in buttons) {
            button.color = bgColor
        }
    }

    override fun act(delta: Float) {
        if (hasFocus) {
            // TODO: use keymapping
            if (Gdx.input.isKeyJustPressed(Input.Keys.ENTER)) {
                pressed(selectedIndex)
            }
            if (Gdx.input.isKeyJustPressed(Input.Keys.LEFT)) {
                moveCursor(-1, 0)
            }
            if (Gdx.input.isKeyJustPressed(Input.Keys.RIGHT)) {
                moveCursor(1, 0)
            }
            if (Gdx.input.isKeyJustPressed(Input.Keys.UP)) {
                moveCursor(0, -1)
            }
            if (Gdx.input.isKeyJustPressed(Input.Keys.DOWN)) {
                moveCursor(0, 1)
            }
        }
        super.act(delta)
    }

    override fun draw(batch: Batch, parentAlpha: Float) {
        val sh: ShapeDrawer = Shapes.instance
        sh.setColor(bgColor)
        sh.filledRectangle(x, y, width, height)
        sh.setColor(fgColor)
        sh.rectangle(x, y, width, height, 3f)
        super.draw(batch, parentAlpha)
    }

    protected abstract fun pressed(id: Int)

    val selectedButton: RetroButton
        get() = buttons[selectedIndex]
}